diff -urN ipac-ng-1.27.org/agents/ipchains/ipchains.c ipac-ng-1.27/agents/ipchains/ipchains.c
--- ipac-ng-1.27.org/agents/ipchains/ipchains.c	2003-09-04 13:39:54.000000000 +0200
+++ ipac-ng-1.27/agents/ipchains/ipchains.c	2003-09-04 13:47:30.000000000 +0200
@@ -511,8 +511,7 @@
                         	return 0;
 
                         if (memcmp(src, "!", 1) == 0) {
-				fprintf(stderr, "Multiple `!' flags not 
-								    allowed");
+				fprintf(stderr, "Multiple `!' flags not allowed");
 				exit(1);
 			}
                         return 1;
